pkg/audit/audit.go

```diff
@@ -0,0 +1,139 @@
+package audit
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/pborman/uuid"
+
+	k8stypes "k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	contentTypeJSON = "application/json"
+)
+
+const (
+	levelNull = iota
+	levelMetadata
+	levelRequest
+	levelRequestResponse
+)
+
+var (
+	bodyMethods = map[string]bool{
+		http.MethodPut:  true,
+		http.MethodPost: true,
+	}
+)
+
+type auditLog struct {
+	log     *log
+	writer  *LogWriter
+	reqBody []byte
+}
+
+type log struct {
+	AuditID           k8stypes.UID `json:"auditID,omitempty"`
+	RequestURI        string       `json:"requestURI,omitempty"`
+	User              *userInfo    `json:"user,omitempty"`
+	Verb              string       `json:"verb,omitempty"`
+	SourceIP          string       `json:"sourceIP,omitempty"`
+	RequestTimestamp  string       `json:"requestTimestamp,omitempty"`
+	ResponseTimestamp string       `json:"responseTimestamp,omitempty"`
+	ResponseCode      int          `json:"responseCode,omitempty"`
+	RequestHeader     http.Header  `json:"requestHeader,omitempty"`
+	ResponseHeader    http.Header  `json:"responseHeader,omitempty"`
+	RequestBody       []byte       `json:"requestBody,omitempty"`
+	ResponseBody      []byte       `json:"responseBody,omitempty"`
+}
+
+type userInfo struct {
+	Name  string `json:"name,omitempty"`
+	Group string `json:"group,omitempty"`
+}
+
+func new(writer *LogWriter, req *http.Request) (*auditLog, error) {
+	auditLog := &auditLog{
+		writer: writer,
+		log: &log{
+			AuditID:          k8stypes.UID(uuid.NewRandom().String()),
+			RequestURI:       req.RequestURI,
+			Verb:             req.Method,
+			SourceIP:         getSourceIPs(req.RemoteAddr),
+			RequestTimestamp: time.Now().Format(time.RFC3339),
+		},
+	}
+
+	contentType := req.Header.Get("Content-Type")
+	if writer.Level >= levelRequest && bodyMethods[req.Method] && contentType == contentTypeJSON {
+		reqBody, err := readBodyWithoutLosingContent(req)
+		if err != nil {
+			return nil, err
+		}
+		auditLog.reqBody = reqBody
+	}
+	return auditLog, nil
+}
+
+func (a *auditLog) write(reqHeaders, resHeaders http.Header, resCode int, resBody []byte) error {
+	userInfo := &userInfo{
+		Name:  reqHeaders.Get("Impersonate-User"),
+		Group: reqHeaders.Get("Impersonate-Group"),
+	}
+	a.log.User = userInfo
+	a.log.ResponseTimestamp = time.Now().Format(time.RFC3339)
+	a.log.RequestHeader = reqHeaders
+	a.log.ResponseHeader = resHeaders
```

Comment:
Don't log Cookie, Set-Cookie headers

Reply:
Do you think we need the headers? The request head and response header are lists below, seems like they are meaningless to most of user.
request header"
```
"requestHeader": {
        "Accept": [
            "application/json"
        ],
        "Accept-Encoding": [
            "gzip, deflate, br"
        ],
        "Accept-Language": [
            "en,zh-CN;q=0.9,zh;q=0.8"
        ],
        "Connection": [
            "keep-alive"
        ],
        "Content-Type": [
            "application/json"
        ],
        "Impersonate-Group": [
            "system:authenticated"
        ],
        "Impersonate-User": [
            "user-bkszq"
        ],
        "Referer": [
            "https://localhost:8443/c/local/nodes/local:machine-r7nfg"
        ],
        "User-Agent": [
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36"
        ],
        "X-Api-Action-Links": [
            "actionLinks"
        ],
        "X-Api-Csrf": [
            "853952e7fc"
        ],
        "X-Api-No-Challenge": [
            "true"
        ]
    },
```

response header:
```
"responseHeader": {
        "Content-Type": [
            "application/json"
        ],
        "Expires": [
            "Wed 24 Feb 1982 18:42:00 GMT"
        ],
        "X-Api-Schemas": [
            "https://localhost:8443/v3/schemas"
        ]
    }
```